### PR TITLE
Implement dynamic inference engine selection #45

### PR DESCRIPTION
**Pull request by itsknk**
For - issue #45
 
Did -  Identify the system type and architecture, and choose the appropriate inference engine accordingly.

How?
- Two new helper functions: `is_apple_silicon()` and `is_intel_mac()`.
- A new conditional block that replaces the previous simple check. This block detects the system type (macOS, Linux, or other) and the Mac architecture (Apple Silicon or Intel) if applicable.
- Print statements to indicate which system type was detected and which inference engine is being used.

### CODE REVIEW

**Comment by AlexCheema** [comment]
Did you commit the helper functions? I don’t see them.

**Comment by itsknk** [comment]
Hey @AlexCheema, please check now.

**Comment by AlexCheema** [comment]
Looks good. Did you manage to test this on an Intel Mac by chance? I don't have one myself. @itsknk

**Comment by itsknk** [comment]
> Looks good. Did you manage to test this on an Intel Mac by chance? I don't have one myself. @itsknk

Yes, @AlexCheema. I ran it mine and it was selecting tinygrad as default.

Here's how it looks:
Screenshot 2024-07-20 at 7 16 51 AM

File: exo/helpers.py [hunk 1]
```diff
@@ -54,6 +84,11 @@ def print_yellow_exo():
     reset = "\033[0m"    # ANSI escape code to reset color
     print(f"{yellow}{exo_text}{reset}")
 
+def print_system_info():
+    system_info = get_system_info()
+    print(f"Detected system: {system_info}")
+    print(f"Using inference engine: {get_inference_engine().__class__.__name__}")
```
Comment (AlexCheema) [inline exo/helpers.py:90]: This is creating a new inference engine which is quite a heavy operation / should only be done once in the app.

File: main.py [hunk 2]
```diff
@@ -27,33 +24,14 @@
...
-        inference_engine = MLXDynamicShardInferenceEngine()
-    elif args.inference_engine == "tinygrad":
-        from exo.inference.tinygrad.inference import TinygradDynamicShardInferenceEngine
-        import tinygrad.helpers
-        tinygrad.helpers.DEBUG.value = int(os.getenv("TINYGRAD_DEBUG", default="0"))
-        inference_engine = TinygradDynamicShardInferenceEngine()
-    else:
-        raise ValueError(f"Inference engine {args.inference_engine} not supported")
-print(f"Using inference engine {inference_engine.__class__.__name__}")
+print_system_info()
```
Comment (AlexCheema) [inline main.py:27]: rather than a separate `print_system_info` just print directly here

Comment (AlexCheema) [reply main.py:27]: and use the `inference_engine` to print its class name here

**Comment by itsknk** [comment]
@AlexCheema, please verify this now. 

Update:
- The `print_system_info()` function has been removed from `helpers.py`.
- The system information and inference engine type are printed directly in `main.py`.
- The inference engine is created only once in `main.py`

File: exo/helpers.py [hunk 3]
```diff
@@ -15,6 +17,34 @@
...
 
+def is_apple_silicon():
+    return psutil.MACOS and platform.machine() == 'arm64'
+
+def is_intel_mac():
+    return psutil.MACOS and platform.machine() in ['x86_64', 'i386']
+
+def get_system_info():
+    if psutil.MACOS:
+        if is_apple_silicon():
```
Comment (AlexCheema) [inline exo/helpers.py:28]: Why are these separate functions but below you check psutil.LINUX.
Remove the functions and stick to one convention

**Comment by AlexCheema** [comment]
Looks good apart from the one comment I made.

**Comment by AlexCheema** [comment]
Merged, good job!

**Pull request merged**